First version of the LTR guide. (#105956)
elastic/elasticsearch: docs/reference/search/search-your-data/learning-to-rank-model-training.asciidoc

[[learning-to-rank-model-training]]
=== Deploy and manage Learning To Rank models
++++
<titleabbrev>Deploy and manage LTR models</titleabbrev>
++++

preview::["The Learning To Rank feature is in technical preview and may be changed or removed in a future release. Elastic will work to fix any issues, but this feature is not subject to the support SLA of official GA features."]

> NOTE: This feature is available for Elastic Stack versions 8.12.0 and newer and requires a https://www.elastic.co/pricing[Platinum subscription] or higher.

Reviewer: Should align these warnings per [this comment](https://github.com/elastic/elasticsearch/pull/105956/files#r1514042897)

[discrete]
[[learning-to-rank-model-training-workflow]]
==== Train and deploy a model using Eland

> https://xgboost.readthedocs.io/en/stable/[XGBoost^] model training typically leverages a standard, Python data science toolkit such as Pandas and scikit-learn

Reviewer: ```suggestion
Typically, the https://xgboost.readthedocs.io/en/stable/[XGBoost^] model training process uses standard Python data science tools like Pandas and scikit-learn.

```

> We have developed an example notebook available https://github.com/elastic/elasticsearch-labs/blob/main/notebooks/search/08-learning-to-rank.ipynb[here^], detailing an end-to-end model training and deployment workflow.

Reviewer: ```suggestion
We have developed an
https://github.com/elastic/elasticsearch-labs/blob/main/notebooks/search/08-learning-to-rank.ipynb[example
notebook^] available in the `elasticsearch-labs` repo. This interactive Python notebook
details an end-to-end model training and deployment workflow.
```

> We highly recommend integrating https://eland.readthedocs.io/[eland^] into your workflow since it provides some important features in the workflow to integrate Learning To Rank in {es}:

Reviewer: ```suggestion
We highly recommend using https://eland.readthedocs.io/[eland^] in your workflow, because it provides important functionalities for working with LTR in {es}. Use eland to:
```

Reviewer: eland or Eland? Just a question of consistency...

* Configure feature extraction

* Extract features for training

> * Deploy the model into {es}

Reviewer: ```suggestion
* Deploy the model in {es}
```

[discrete]
[[learning-to-rank-model-training-feature-definition]]
===== Configure feature extraction in Eland

> Feature extractors are defined using templated queries. https://eland.readthedocs.io/[Eland^] provides the `eland.ml.ltr.QueryFeatureExtractor` to define these feature extractors directly in Python:

Reviewer: If we want to backport to 8.12 it would make sense for the Eland docs to be up to date, otherwise users won't find any LTR info there. cc @pquentin

Author: I would like to be able to publish this guide for 8.12 and fix the eland doc in parallel. @pquentin is it possible?

Reviewer: Another option is to merge to 8.13 and main, and wait to manually backport to 8.12 until those Eland docs are up to date

Author: I would prefer to merge right now to 8.12.0 because some customer of us are currently invited to test it.

[source,python]
----
from eland.ml.ltr import QueryFeatureExtractor

feature_extractors=[
    # We want to use the score of the match query for the title field as a features:
    QueryFeatureExtractor(
        feature_name="title_bm25",
        query={"match": {"title": "{{query}}"}}
    ),
    # We can use a script_score query to get the value of the field rating directly as a feature:
    QueryFeatureExtractor(
        feature_name="popularity",
        query={
            "script_score": {
                "query": {"exists": {"field": "popularity"}},
                "script": {"source": "return doc['popularity'].value;"},
            }
        },
    ),
    # We can execute a script on the value of the query and use the return value as a feature:
    QueryFeatureExtractor(
        feature_name="query_length",
        query={
            "script_score": {
                "query": {"match_all": {}},
                "script": {
                    "source": "return params['query'].splitOnToken(' ').length;",
                    "params": {
                        "query": "{{query}}",
                    }
                },
            }
        },
    ),
]
----
// NOTCONSOLE

> Once the feature extractors have been defined, they are wrapped within an `eland.ml.ltr.LTRModelConfig` object which will be used in subsequent steps of the training process:

Reviewer: ```suggestion
Once the feature extractors have been defined, they are wrapped in an `eland.ml.ltr.LTRModelConfig` object for use in later training steps:
```

[source,python]
----
from eland.ml.ltr import LTRModelConfig

ltr_config = LTRModelConfig(feature_extractors)
----
// NOTCONSOLE

[discrete]
[[learning-to-rank-model-training-feature-extraction]]
===== Extracting features for training

> One of the most important steps of the training process is to build the dataset that will be used by extracting and adding features to it. Eland provides another helper class, `eland.ml.ltr.FeatureLogger`, to aid in this process:

Reviewer: ```suggestion
Building your dataset is a critical step in the training process. This involves
extracting relevant features and adding them to your judgment list. We
recommend using Eland's `eland.ml.ltr.FeatureLogger` helper class for this
process.
```

[source,python]
----
from eland.ml.ltr import FeatureLogger

# Create a feature logger that will be used to query {es} to retrieve the features:
feature_logger = FeatureLogger(es_client, MOVIE_INDEX, ltr_config)
----
// NOTCONSOLE

> The FeatureLogger provides an `extract_features` method allowing you to extract features for a list of specific documents from your judgment list. At the same time, query parameters used by the feature extractors defined earlier can be passed:

Reviewer: ```suggestion
The FeatureLogger provides an `extract_features` method which enables you to extract features for a list of specific documents from your judgment list. At the same time, you can pass query parameters to the feature extractors defined earlier:
```

Reviewer: active voice reads clearer

[source,python]
----
feature_logger.extract_features(
    query_params:{"query": "foo"},
    doc_ids=["doc-1", "doc-2"]
)
----
// NOTCONSOLE

> Our https://github.com/elastic/elasticsearch-labs/blob/main/notebooks/search/08-learning-to-rank.ipynb[example notebook^] provides a complete example explaining how to use the `FeatureLogger` to add features to the judgment list in order to build the training dataset.

Reviewer: ```suggestion
Our https://github.com/elastic/elasticsearch-labs/blob/main/notebooks/search/08-learning-to-rank.ipynb[example notebook^] explains how to use the `FeatureLogger` to build a training dataset, by adding features to a judgment list.
```

[discrete]
[[learning-to-rank-model-training-feature-extraction-notes]]

> ====== Notes on features extraction

Reviewer: ```suggestion
====== Notes on feature extraction
```

> * We strongly advise against implementing feature extraction on your own. It's crucial to maintain consistency in feature extraction between the training environment and inference in {es}. By utilizing eland tooling, which is developed and tested in tandem with {es}, you can ensure that they function together consistently.

Reviewer: ```suggestion
* We strongly advise against implementing feature extraction on your own. It's crucial to maintain consistency in feature extraction between the training environment and inference in {es}. By using eland tooling, which is developed and tested in tandem with {es}, you can ensure that they function together consistently.
```

nit: I always replace utilize with use where possible :)

> * Feature extraction is performed by executing queries on the {es} server which could cause a lot of stress on your cluster, especially when your judgment list contains a lot of examples or you have many features. Our feature logger implementation is designed to minimize the number of search requests sent to the server in order to reduce the load, however building the training dataset might best be performed using an {es} cluster that is isolated from any user-facing, production traffic

Reviewer: ```suggestion
* Feature extraction is performed by executing queries on the {es} server. This could put a lot of stress on your cluster, especially when your judgment list contains a lot of examples or you have many features. Our feature logger implementation is designed to minimize the number of search requests sent to the server and reduce load. However, it might be best to build your training dataset using an {es} cluster that is isolated from any user-facing, production traffic.
```

Reviewer: shorter sentences for readability

[discrete]
[[learning-to-rank-model-deployment]]
===== Deploy your model into {es}

> Once your model is trained you will be able to deploy it into your {es} cluster. For this purpose, eland provides the `MLModel.import_ltr_model method`:

Reviewer: ```suggestion
Once your model is trained you will be able to deploy it in your {es} cluster. You can use Eland's `MLModel.import_ltr_model method`:
```

[source,python]
----
from eland.ml import MLModel

LEARNING_TO_RANK_MODEL_ID="ltr-model-xgboost"

MLModel.import_ltr_model(
    es_client=es_client,
    model=ranker,
    model_id=LEARNING_TO_RANK_MODEL_ID,
    ltr_model_config=ltr_config,
    es_if_exists="replace",
)
----
// NOTCONSOLE

> This method will serialize the trained model and the Learning To Rank configuration (including feature extraction) in a format that {es} can understand before sending it to Elasticsearch using the <<put-trained-models, Create Trained Models API>>.

Reviewer: ```suggestion
This method will serialize the trained model and the Learning To Rank configuration (including feature extraction) in a format that {es} can understand. The model is then deployed to {es} using the <<put-trained-models, Create Trained Models API>>.
```

> This method will serialize the trained model and the Learning To Rank configuration (including feature extraction) in a format that {es} can understand before sending it to Elasticsearch using the https://www.elastic.co/guide/en/elasticsearch/reference/current/put-trained-models.htmlp[Create Trained Models API].

Reviewer: ```suggestion
This method will serialize the trained model and the Learning To Rank configuration (including feature extraction) in a format that {es} can understand before sending it to Elasticsearch using the <<put-trained-models,Create Trained Models API>>.
```

Reviewer: `.htmlp` is breaking CI, but we can also change this to internal link formatting

> The following types of models are supported for Learning To Rank: https://scikit-learn.org/stable/modules/generated/sklearn.tree.DecisionTreeRegressor.html[`DecisionTreeRegressor`^],  https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestRegressor.html[`RandomForestRegressor`^], https://lightgbm.readthedocs.io/en/latest/pythonapi/lightgbm.LGBMRegressor.html[`LGBMRegressor`^], https://xgboost.readthedocs.io/en/stable/python/python_api.html#xgboost.XGBRanker[`XGBRanker`^], https://xgboost.readthedocs.io/en/stable/python/python_api.html#xgboost.XGBRegressor[`XGBRegressor`^].

Reviewer: ```suggestion
* https://scikit-learn.org/stable/modules/generated/sklearn.tree.DecisionTreeRegressor.html[`DecisionTreeRegressor`^]
* https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestRegressor.html[`RandomForestRegressor`^]
* https://lightgbm.readthedocs.io/en/latest/pythonapi/lightgbm.LGBMRegressor.html[`LGBMRegressor`^]
* https://xgboost.readthedocs.io/en/stable/python/python_api.html#xgboost.XGBRanker[`XGBRanker`^]
* https://xgboost.readthedocs.io/en/stable/python/python_api.html#xgboost.XGBRegressor[`XGBRegressor`^]

```

Reviewer: bulleted list reads nicer

More model types will be supported in the future.

[discrete]
[[learning-to-rank-model-management]]
==== Learning To Rank model management

> Once your model is deployed into {es} it is possible to manage it using the https://www.elastic.co/guide/en/elasticsearch/reference/current/ml-df-trained-models-apis.html[trained model APIs].

Reviewer: ```suggestion
Once your model is deployed in {es} you can manage it using the https://www.elastic.co/guide/en/elasticsearch/reference/current/ml-df-trained-models-apis.html[trained model APIs].
You're now ready to work with your LTR model as a rescorer at <<learning-to-rank-search-usage, search time>>.
```

> Once your model is deployed into {es} it is possible to manage it using the https://www.elastic.co/guide/en/elasticsearch/reference/current/ml-df-trained-models-apis.html[trained model APIs].

Reviewer: ```suggestion
Once your model is deployed into {es} it is possible to manage it using the <<ml-df-trained-models-apis,trained model APIs>>.
```

Author: Fixed!!!

[discrete]
[[learning-to-rank-model-deployment]]
===== Deploy your model into {es}

Once your model is trained you will be able to deploy it in your {es} cluster. You can use Eland's `MLModel.import_ltr_model method`:

[source,python]
----
from eland.ml import MLModel

LEARNING_TO_RANK_MODEL_ID="ltr-model-xgboost"

MLModel.import_ltr_model(
    es_client=es_client,
    model=ranker,
    model_id=LEARNING_TO_RANK_MODEL_ID,
    ltr_model_config=ltr_config,
    es_if_exists="replace",
)
----
// NOTCONSOLE

This method will serialize the trained model and the Learning To Rank configuration (including feature extraction) in a format that {es} can understand. The model is then deployed to {es} using the <<put-trained-models, Create Trained Models API>>.

> * https://scikit-learn.org/stable/modules/generated/sklearn.tree.DecisionTreeRegressor.html[`DecisionTreeRegressor`^]

Reviewer: ```suggestion
The following types of models are currently supported for LTR with {es}:

* https://scikit-learn.org/stable/modules/generated/sklearn.tree.DecisionTreeRegressor.html[`DecisionTreeRegressor`^]
```

[[learning-to-rank-model-training]]

> === Deploy and manage Learning To Rank models

Author: There is a lot of other place where we are using this capitalization for LTR, so it seems to be consistent

++++
<titleabbrev>Deploy and manage LTR models</titleabbrev>
++++

> preview::["The Learning To Rank feature is in technical preview and may be changed or removed in a future release. Elastic will work to fix any issues, but this feature is not subject to the support SLA of official GA features."]

Reviewer 2: Any reason we don't want to just use the default copy here that comes with `preview::[]`?

Author: Just wanted to put the emphasis on the LTR feature.

We have developed an
https://github.com/elastic/elasticsearch-labs/blob/main/notebooks/search/08-learning-to-rank.ipynb[example
notebook^] available in the `elasticsearch-labs` repo. This interactive Python notebook
details an end-to-end model training and deployment workflow.

We highly recommend using https://eland.readthedocs.io/[eland^] in your workflow, because it provides important functionalities for working with LTR in {es}. Use eland to:

* Configure feature extraction

* Extract features for training

* Deploy the model in {es}

[discrete]
[[learning-to-rank-model-training-feature-definition]]
===== Configure feature extraction in Eland

Feature extractors are defined using templated queries. https://eland.readthedocs.io/[Eland^] provides the `eland.ml.ltr.QueryFeatureExtractor` to define these feature extractors directly in Python:

[source,python]
----
from eland.ml.ltr import QueryFeatureExtractor

feature_extractors=[
    # We want to use the score of the match query for the title field as a features:
    QueryFeatureExtractor(
        feature_name="title_bm25",
        query={"match": {"title": "{{query}}"}}
    ),

>     # We can use a script_score query to get the value of the field rating directly as a feature:

Reviewer 2: Consider shortening this comment or breaking into 2 lines so the code snippet window doesn't scroll.

Building your dataset is a critical step in the training process. This involves
extracting relevant features and adding them to your judgment list. We
recommend using Eland's `eland.ml.ltr.FeatureLogger` helper class for this
process.

[source,python]
----
from eland.ml.ltr import FeatureLogger

# Create a feature logger that will be used to query {es} to retrieve the features:
feature_logger = FeatureLogger(es_client, MOVIE_INDEX, ltr_config)
----
// NOTCONSOLE

The FeatureLogger provides an `extract_features` method which enables you to extract features for a list of specific documents from your judgment list. At the same time, you can pass query parameters to the feature extractors defined earlier:

[source,python]
----
feature_logger.extract_features(
    query_params:{"query": "foo"},
    doc_ids=["doc-1", "doc-2"]
)
----
// NOTCONSOLE

Our https://github.com/elastic/elasticsearch-labs/blob/main/notebooks/search/08-learning-to-rank.ipynb[example notebook^] explains how to use the `FeatureLogger` to build a training dataset, by adding features to a judgment list.

[discrete]
[[learning-to-rank-model-training-feature-extraction-notes]]
====== Notes on feature extraction

> * We strongly advise against implementing feature extraction on your own. It's crucial to maintain consistency in feature extraction between the training environment and inference in {es}. By using eland tooling, which is developed and tested in tandem with {es}, you can ensure that they function together consistently.

Reviewer 2: Should this be a warning?

Author: The warning presentation does not fit well with the other notes. I prefer to keep this into a note.
